### PR TITLE
Port CodeGenPrepare to new pass manager

### PR DESCRIPTION
**Pull request by Krishna-13-cyber**

### CODE REVIEW

File: llvm/lib/CodeGen/CodeGenPrepare.cpp [hunk 1]
```diff
@@ -486,72 +432,185 @@ class CodeGenPrepare : public FunctionPass {
...
+
+  auto *TM = &TPC->getTM<TargetMachine>();
+  auto &TTI = getAnalysis<TargetTransformInfoWrapperPass>().getTTI(F);
+  auto &LI = getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
+  auto &PSI = getAnalysis<ProfileSummaryInfoWrapperPass>().getPSI();
+  auto &TLInfo = getAnalysis<TargetLibraryInfoWrapperPass>().getTLI(F);
+  // auto &BBSectionsProfileReader =
+  //     getAnalysisIfAvailable<BasicBlockSectionsProfileReader>();
+
+  CodeGenPrepare CGP;
```
Comment (arsenm) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:542]: I'd expect this to be constructed with the TargetMachine, not have a TM argument

File: llvm/lib/CodeGen/CodeGenPrepare.cpp [hunk 2]
```diff
@@ -486,72 +432,185 @@ class CodeGenPrepare : public FunctionPass {
...
+  auto *TPC = getAnalysisIfAvailable<TargetPassConfig>();
+  if (!TPC)
+    return false;
+
+  auto *TM = &TPC->getTM<TargetMachine>();
+  auto &TTI = getAnalysis<TargetTransformInfoWrapperPass>().getTTI(F);
+  auto &LI = getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
+  auto &PSI = getAnalysis<ProfileSummaryInfoWrapperPass>().getPSI();
+  auto &TLInfo = getAnalysis<TargetLibraryInfoWrapperPass>().getTLI(F);
+  // auto &BBSectionsProfileReader =
```
Comment (arsenm) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:539]: Don't comment this out?

File: llvm/lib/CodeGen/CodeGenPrepare.cpp [hunk 3]
```diff
@@ -486,72 +432,185 @@ class CodeGenPrepare : public FunctionPass {
...
-FunctionPass *llvm::createCodeGenPreparePass() { return new CodeGenPrepare(); }
+FunctionPass *llvm::createCodeGenPreparePass() {
+ return new CodeGenPrepareLegacy();
+}
 
-bool CodeGenPrepare::runOnFunction(Function &F) {
+bool CodeGenPrepareLegacy::runOnFunction(Function &F) {
   if (skipFunction(F))
     return false;
+  auto *TPC = getAnalysisIfAvailable<TargetPassConfig>();
```
Comment (arsenm) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:530]: Do you still need this if the TM is passed to the constructor function?

File: llvm/lib/CodeGen/CodeGenPrepare.cpp [hunk 4]
```diff
@@ -486,72 +432,185 @@ class CodeGenPrepare : public FunctionPass {
...
+  CodeGenPrepare CGP;
+  return CGP.run(F, TM, TTI, LI, PSI, TLInfo);
+}
+
+PreservedAnalyses CodeGenPreparePass::run(Function &F, FunctionAnalysisManager &AM) {
+ auto &TTI = AM.getResult<TargetIRAnalysis>(F);
+ auto &LI = AM.getResult<LoopAnalysis>(F);
+ auto &PSI = AM.getResult<ProfileSummaryAnalysis>(F);
+ auto &TLInfo = AM.getResult<TargetLibraryAnalysis>(F);
+// auto &BasicBlockSectionsProfileReader = AM.getResult<BasicBlockSectionsProfileReader>(F);
```
Comment (arsenm) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:551]: This is another analysis pass that needs to be ported as a dependency

File: llvm/lib/CodeGen/CodeGenPrepare.cpp [hunk 5]
```diff
@@ -486,72 +432,185 @@ class CodeGenPrepare : public FunctionPass {
...
+                            const TargetLibraryInfo &TLInfo) {
+  if (DisableBranchOpts)
+    return false;
 
   DL = &F.getParent()->getDataLayout();
 
   bool EverMadeChange = false;
 
-  TM = &getAnalysis<TargetPassConfig>().getTM<TargetMachine>();
+  // TM = &getAnalysisUsage<TargetPassConfig>().getTM<TargetMachine>();
```
Comment (arsenm) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:574]: Drop this, it comes from the pass constructor

File: llvm/lib/CodeGen/CodeGenPrepare.cpp [hunk 6]
```diff
@@ -486,72 +432,185 @@ class CodeGenPrepare : public FunctionPass {
...
+ bool Changed = CGP.run(F, TM, TTI, LI, PSI, TLInfo);
+ if (!Changed)
+   return PreservedAnalyses::all();
+ PreservedAnalyses PA;
+ PA.preserveSet<CFGAnalyses>();
+ PA.preserve<LoopAnalysis>();
+ return PA;
+}
+
+bool CodeGenPrepare::run(Function &F, const TargetMachine *TM,
```
Comment (arsenm) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:563]: Don't need the TM argument

File: llvm/lib/CodeGen/CodeGenPrepare.cpp [hunk 7]
```diff
@@ -486,72 +432,185 @@ class CodeGenPrepare : public FunctionPass {
...
-  TLInfo = &getAnalysis<TargetLibraryInfoWrapperPass>().getTLI(F);
-  TTI = &getAnalysis<TargetTransformInfoWrapperPass>().getTTI(F);
-  LI = &getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
-  BPI.reset(new BranchProbabilityInfo(F, *LI));
-  BFI.reset(new BlockFrequencyInfo(F, *BPI, *LI));
-  PSI = &getAnalysis<ProfileSummaryInfoWrapperPass>().getPSI();
-  BBSectionsProfileReader =
-      getAnalysisIfAvailable<BasicBlockSectionsProfileReader>();
+  // TLInfo = &getAnalysis<TargetLibraryInfoWrapperPass>().getTLI(F);
+  // TTI = &getAnalysis<TargetTransformInfoWrapperPass>().getTTI(F);
```
Comment (arsenm) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:579]: Get TTI with &AM.getResult<TargetIRAnalysis>(F)

File: llvm/lib/CodeGen/CodeGenPrepare.cpp [hunk 8]
```diff
@@ -486,72 +432,185 @@ class CodeGenPrepare : public FunctionPass {
...
   TRI = SubtargetInfo->getRegisterInfo();
-  TLInfo = &getAnalysis<TargetLibraryInfoWrapperPass>().getTLI(F);
-  TTI = &getAnalysis<TargetTransformInfoWrapperPass>().getTTI(F);
-  LI = &getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
-  BPI.reset(new BranchProbabilityInfo(F, *LI));
-  BFI.reset(new BlockFrequencyInfo(F, *BPI, *LI));
-  PSI = &getAnalysis<ProfileSummaryInfoWrapperPass>().getPSI();
-  BBSectionsProfileReader =
-      getAnalysisIfAvailable<BasicBlockSectionsProfileReader>();
+  // TLInfo = &getAnalysis<TargetLibraryInfoWrapperPass>().getTLI(F);
```
Comment (arsenm) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:578]: AM.getResult<TargetLibraryAnalysis>(F)

File: llvm/lib/CodeGen/CodeGenPrepare.cpp [hunk 9]
```diff
@@ -486,72 +432,185 @@ class CodeGenPrepare : public FunctionPass {
...
+  // TTI = &getAnalysis<TargetTransformInfoWrapperPass>().getTTI(F);
+  // LI = &getAnalysis<LoopInfoWrapperPass>().getLoopInfo();
```
Comment (arsenm) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:580]: AM.getResult<LoopAnalysis>(F)

File: llvm/lib/CodeGen/CodeGenPrepare.cpp [hunk 10]
```diff
@@ -486,72 +432,185 @@ class CodeGenPrepare : public FunctionPass {
...
+  BPI.reset(new BranchProbabilityInfo(F, LI));
+  BFI.reset(new BlockFrequencyInfo(F, *BPI, LI));
+  // PSI = &getAnalysis<ProfileSummaryInfoWrapperPass>().getPSI();
```
Comment (arsenm) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:583]: Looking at ConstantHoisting as an example, I think you need something like:

```
  auto &MAMProxy = AM.getResult<ModuleAnalysisManagerFunctionProxy>(F);
  auto *PSI = MAMProxy.getCachedResult<ProfileSummaryAnalysis>(*F.getParent());
```

File: llvm/lib/CodeGen/CodeGenPrepare.cpp [hunk 11]
```diff
@@ -486,72 +432,185 @@ class CodeGenPrepare : public FunctionPass {
...
+  // BBSectionsProfileReader =
+  //     getAnalysisIfAvailable<BasicBlockSectionsProfileReader>();
```
Comment (arsenm) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:585]: This needs to be ported for access through the analysis manager

**Review by arsenm — CHANGES_REQUESTED** [review]

**Comment by Krishna-13-cyber** [comment]
Hi @nickleus27 ,
Thanks for carrying the work as I couldn't continue due to hardware constraints on my side.

Closing this PR as done in #75380

**Pull request closed without merging**